runtime/vm/BytecodeInterpreter.hpp

```diff
@@ -1195,10 +1195,43 @@ class INTERPRETER_CLASS
 					_currentThread->tempSlot = (UDATA)invalidIndex;
 					rc = THROW_AIOB;
 				} else {
-					I_32 value = VM_ArrayCopyHelpers::referenceArrayCopy(_currentThread, srcObject, srcStart, destObject, destStart, elementCount);
-					if (-1 != value) {
-						buildInternalNativeStackFrame(REGISTER_ARGS);
-						rc = THROW_ARRAY_STORE;
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+					J9Class *srcClazz = J9OBJECT_CLAZZ(_currentThread, srcObject);
+					J9Class *destClazz = J9OBJECT_CLAZZ(_currentThread, destObject);
+					J9Class *srcComponentClass = ((J9ArrayClass *)srcClazz)->componentType;
+					J9Class *destComponentClass = ((J9ArrayClass *)destClazz)->componentType;
+
+					if (J9_IS_J9CLASS_FLATTENED(srcClazz) && J9_IS_J9CLASS_FLATTENED(destClazz) && !J9_ARE_ANY_BITS_SET(srcComponentClass->classFlags, J9ClassHasReferences) && !J9_ARE_ANY_BITS_SET(destComponentClass->classFlags, J9ClassHasReferences)) {
```

Comment:
sorry, I should have suggested J9_ARE_NO_BITS_SET(...

Reply:
updated in [e2d43b1](https://github.com/eclipse-openj9/openj9/commit/e2d43b16d918f6ebc2210f113be6b9ff931db4ce)